<a href="https://colab.research.google.com/github/OlhaYastrebova/goit-ds-hw/blob/main/Hw_13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import tensorflow as tf
from keras import layers, models
from keras.datasets import fashion_mnist
from keras.utils import to_categorical

# Завантажуємо дані
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Нормалізуємо дані
train_images = train_images / 255.0
test_images = test_images / 255.0

# One-hot encoding міток
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)


29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [ ]:
from keras import Input

model = models.Sequential()

# Вхідний шар
model.add(Input(shape=(28, 28, 1)))

# Шар згортки
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Додаткові шари згортки
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Перетворення до одного виміру
model.add(layers.Flatten())

# Щільний шар
model.add(layers.Dense(128, activation='relu'))

# Вихідний шар
model.add(layers.Dense(10, activation='softmax'))

# Компіляція моделі
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])


In [ ]:
# Тренуємо модель
history = model.fit(train_images, train_labels, epochs=10, batch_size=64, validation_split=0.2)

# Оцінюємо модель
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Точність на тестових даних: {test_acc}')


Epoch 1/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 50s 64ms/step - accuracy: 0.6974 - loss: 0.8229 - val_accuracy: 0.8462 - val_loss: 0.4140
Epoch 2/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 83s 66ms/step - accuracy: 0.8643 - loss: 0.3738 - val_accuracy: 0.8695 - val_loss: 0.3515
Epoch 3/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 78s 60ms/step - accuracy: 0.8868 - loss: 0.3096 - val_accuracy: 0.8879 - val_loss: 0.3115
Epoch 4/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 60ms/step - accuracy: 0.8970 - loss: 0.2756 - val_accuracy: 0.8986 - val_loss: 0.2828
Epoch 5/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 84s 63ms/step - accuracy: 0.9098 - loss: 0.2479 - val_accuracy: 0.8999 - val_loss: 0.2767
Epoch 6/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 46s 61ms/step - accuracy: 0.9188 - loss: 0.2207 - val_accuracy: 0.9076 - val_loss: 0.2580
Epoch 7/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 81s 60ms/step - accuracy: 0.9244 - loss: 0.2047 - val_accuracy: 0.9089 - val_loss: 0.2718
Epoch 8/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 60ms/step - accuracy: 0.9306 - loss: 0.1885 - 

In [ ]:
from sklearn.metrics import classification_report
import numpy as np

# Прогнозування на тестових даних
y_pred = model.predict(test_images)

# Перетворюємо one-hot encoded мітки на звичайні
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(test_labels, axis=1)

# Виведення classification_report
report = classification_report(y_true_classes, y_pred_classes, target_names=[
    'T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
    'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'])
print(report)


313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 14ms/step
              precision    recall  f1-score   support

 T-shirt/top       0.87      0.86      0.86      1000
     Trouser       1.00      0.97      0.99      1000
    Pullover       0.83      0.90      0.87      1000
       Dress       0.92      0.91      0.91      1000
        Coat       0.87      0.84      0.85      1000
      Sandal       0.99      0.98      0.98      1000
       Shirt       0.75      0.74      0.74      1000
     Sneaker       0.95      0.97      0.96      1000
         Bag       0.97      0.98      0.98      1000
  Ankle boot       0.97      0.96      0.97      1000

    accuracy                           0.91     10000
   macro avg       0.91      0.91      0.91     10000
weighted avg       0.91      0.91      0.91     10000



Додавання регуляризації

In [ ]:
# Побудова моделі
model = models.Sequential()

# Вхідний шар
model.add(layers.Input(shape=(28, 28, 1)))

# Шар згортки
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Dropout(0.3))  # Додаємо Dropout після шару MaxPooling

# Додаткові шари згортки
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Dropout(0.3))  # Додаємо Dropout після шару MaxPooling

model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Перетворення до одного виміру
model.add(layers.Flatten())

# Щільний шар з регуляризацією (Dropout)
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dropout(0.4))  # Dropout після Dense шару

# Вихідний шар
model.add(layers.Dense(10, activation='softmax'))

# Компіляція моделі
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Тренування моделі
history = model.fit(train_images, train_labels, epochs=10, batch_size=64, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Точність на тестових даних: {test_acc}')


Epoch 1/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 58s 73ms/step - accuracy: 0.6332 - loss: 1.0022 - val_accuracy: 0.8390 - val_loss: 0.4345
Epoch 2/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 77s 67ms/step - accuracy: 0.8277 - loss: 0.4727 - val_accuracy: 0.8684 - val_loss: 0.3564
Epoch 3/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 67ms/step - accuracy: 0.8543 - loss: 0.3994 - val_accuracy: 0.8639 - val_loss: 0.3629
Epoch 4/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 49s 65ms/step - accuracy: 0.8673 - loss: 0.3653 - val_accuracy: 0.8915 - val_loss: 0.2982
Epoch 5/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 84s 67ms/step - accuracy: 0.8780 - loss: 0.3384 - val_accuracy: 0.8918 - val_loss: 0.2892
Epoch 6/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 49s 65ms/step - accuracy: 0.8823 - loss: 0.3203 - val_accuracy: 0.8923 - val_loss: 0.2852
Epoch 7/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 66ms/step - accuracy: 0.8903 - loss: 0.3028 - val_accuracy: 0.8993 - val_loss: 0.2693
Epoch 8/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 66ms/step - accuracy: 0.8894 - loss: 0.2982 - 

In [ ]:
from sklearn.metrics import classification_report
import numpy as np

# Прогнозування на тестових даних
y_pred = model.predict(test_images)

# Перетворюємо one-hot encoded мітки на звичайні
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(test_labels, axis=1)

# Виведення classification_report
report = classification_report(y_true_classes, y_pred_classes, target_names=[
    'T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
    'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'])
print(report)

313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 13ms/step
              precision    recall  f1-score   support

 T-shirt/top       0.86      0.85      0.86      1000
     Trouser       0.99      0.98      0.98      1000
    Pullover       0.86      0.86      0.86      1000
       Dress       0.92      0.89      0.91      1000
        Coat       0.84      0.84      0.84      1000
      Sandal       0.99      0.97      0.98      1000
       Shirt       0.69      0.73      0.71      1000
     Sneaker       0.93      0.98      0.96      1000
         Bag       0.98      0.98      0.98      1000
  Ankle boot       0.98      0.94      0.96      1000

    accuracy                           0.90     10000
   macro avg       0.90      0.90      0.90     10000
weighted avg       0.90      0.90      0.90     10000



Зміна кількості шарів, нейронів та епох

In [ ]:
# Побудова моделі
model = models.Sequential()

# Вхідний шар
model.add(layers.Input(shape=(28, 28, 1)))

# Шар згортки
model.add(layers.Conv2D(32, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Додаткові шари згортки
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Перетворення до одного виміру
model.add(layers.Flatten())

# Щільні шари
model.add(layers.Dense(256, activation='relu'))  # Додаємо Dense шар із 256 нейронами
model.add(layers.Dense(128, activation='relu'))  # Додаємо ще один Dense шар із 128 нейронами

# Вихідний шар
model.add(layers.Dense(10, activation='softmax'))

# Компіляція моделі
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Тренування моделі
history = model.fit(train_images, train_labels, epochs=20, batch_size=64, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Точність на тестових даних: {test_acc}')

Epoch 1/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 56s 72ms/step - accuracy: 0.6942 - loss: 0.8206 - val_accuracy: 0.8470 - val_loss: 0.4209
Epoch 2/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 77s 65ms/step - accuracy: 0.8643 - loss: 0.3686 - val_accuracy: 0.8791 - val_loss: 0.3311
Epoch 3/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 82s 65ms/step - accuracy: 0.8914 - loss: 0.2929 - val_accuracy: 0.8907 - val_loss: 0.3107
Epoch 4/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 80s 63ms/step - accuracy: 0.9019 - loss: 0.2639 - val_accuracy: 0.8944 - val_loss: 0.2892
Epoch 5/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 49s 66ms/step - accuracy: 0.9120 - loss: 0.2358 - val_accuracy: 0.9018 - val_loss: 0.2689
Epoch 6/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 47s 62ms/step - accuracy: 0.9230 - loss: 0.2104 - val_accuracy: 0.8986 - val_loss: 0.2872
Epoch 7/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 83s 64ms/step - accuracy: 0.9303 - loss: 0.1902 - val_accuracy: 0.9081 - val_loss: 0.2597
Epoch 8/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 81s 63ms/step - accuracy: 0.9365 - loss: 0.1719 - 

In [ ]:
from sklearn.metrics import classification_report
import numpy as np

# Прогнозування на тестових даних
y_pred = model.predict(test_images)

# Перетворюємо one-hot encoded мітки на звичайні
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(test_labels, axis=1)

# Виведення classification_report
report = classification_report(y_true_classes, y_pred_classes, target_names=[
    'T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
    'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'])
print(report)

313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step
              precision    recall  f1-score   support

 T-shirt/top       0.83      0.87      0.85      1000
     Trouser       0.99      0.97      0.98      1000
    Pullover       0.83      0.88      0.85      1000
       Dress       0.89      0.93      0.91      1000
        Coat       0.88      0.81      0.84      1000
      Sandal       0.97      0.98      0.98      1000
       Shirt       0.76      0.68      0.72      1000
     Sneaker       0.93      0.98      0.95      1000
         Bag       0.95      0.98      0.97      1000
  Ankle boot       0.99      0.93      0.96      1000

    accuracy                           0.90     10000
   macro avg       0.90      0.90      0.90     10000
weighted avg       0.90      0.90      0.90     10000



In [ ]:
from keras.optimizers import RMSprop

model.compile(optimizer=RMSprop(learning_rate=0.001),
              loss='categorical_crossentropy',
              metrics=['accuracy'])
# Тренування моделі
history = model.fit(train_images, train_labels, epochs=20, batch_size=64, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Точність на тестових даних: {test_acc}')

Epoch 1/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 48s 62ms/step - accuracy: 0.9833 - loss: 0.0486 - val_accuracy: 0.9096 - val_loss: 0.5352
Epoch 2/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 81s 60ms/step - accuracy: 0.9863 - loss: 0.0368 - val_accuracy: 0.9071 - val_loss: 0.5771
Epoch 3/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 50s 67ms/step - accuracy: 0.9874 - loss: 0.0351 - val_accuracy: 0.9082 - val_loss: 0.5851
Epoch 4/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 77s 61ms/step - accuracy: 0.9879 - loss: 0.0354 - val_accuracy: 0.9060 - val_loss: 0.5926
Epoch 5/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 81s 59ms/step - accuracy: 0.9875 - loss: 0.0340 - val_accuracy: 0.9077 - val_loss: 0.5984
Epoch 6/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 46s 62ms/step - accuracy: 0.9886 - loss: 0.0317 - val_accuracy: 0.9046 - val_loss: 0.6445
Epoch 7/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 80s 59ms/step - accuracy: 0.9895 - loss: 0.0308 - val_accuracy: 0.9087 - val_loss: 0.6871
Epoch 8/20
750/750 ━━━━━━━━━━━━━━━━━━━━ 85s 63ms/step - accuracy: 0.9898 - loss: 0.0289 - 

In [ ]:
from sklearn.metrics import classification_report
import numpy as np

# Прогнозування на тестових даних
y_pred = model.predict(test_images)

# Перетворюємо one-hot encoded мітки на звичайні
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(test_labels, axis=1)

# Виведення classification_report
report = classification_report(y_true_classes, y_pred_classes, target_names=[
    'T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
    'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot'])
print(report)


313/313 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step
              precision    recall  f1-score   support

 T-shirt/top       0.85      0.85      0.85      1000
     Trouser       0.99      0.98      0.99      1000
    Pullover       0.82      0.89      0.85      1000
       Dress       0.89      0.93      0.91      1000
        Coat       0.87      0.86      0.86      1000
      Sandal       0.97      0.98      0.98      1000
       Shirt       0.77      0.70      0.73      1000
     Sneaker       0.94      0.96      0.95      1000
         Bag       0.99      0.97      0.98      1000
  Ankle boot       0.97      0.95      0.96      1000

    accuracy                           0.91     10000
   macro avg       0.91      0.91      0.91     10000
weighted avg       0.91      0.91      0.91     10000



Висновок. Перша і остання моделі дають достатню точність - 91 %.

Згорткова нейронна мережа (CNN) значно перевершує багатошарову мережу (MLP) у задачах класифікації зображень, оскільки вона краще працює з просторовими структурами зображень та може ефективніше розрізняти схожі класи.

MLP має свої переваги у швидкості навчання і простоті, проте для класифікації зображень CNN є кращим вибором.
Зважаючи на це, CNN є більш підходящою архітектурою для класифікації даних типу Fashion MNIST.

In [3]:
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models, optimizers
from tensorflow.keras.applications import VGG16
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.utils import to_categorical

In [4]:
# Завантаження даних (приклад)
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

# Нормалізація даних
x_train = x_train / 255.0
x_test = x_test / 255.0

# Розширення розмірності для зображень в один канал
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)

In [5]:
# Зміна розміру зображень до 32x32
x_train = tf.image.resize(x_train, (32, 32))
x_test = tf.image.resize(x_test, (32, 32))

# Додавання третього каналу (для RGB)
x_train = tf.concat([x_train] * 3, axis=-1)
x_test = tf.concat([x_test] * 3, axis=-1)

In [13]:
# Завантаження VGG16 без класифікаційної частини (тільки згорткові шари)
base_model = VGG16(input_shape=(224, 224, 3), include_top=False, weights='imagenet')
base_model.trainable = False  # Заморожуємо базову модель

In [7]:
from tensorflow.keras.layers import Input, Conv2D, MaxPooling2D, Flatten, Dense
from tensorflow.keras.models import Sequential
# Define the input shape
input_shape = (32, 32, 3)

# Create an input layer
inputs = Input(shape=input_shape)

# Build your model using Sequential
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=None),
    MaxPooling2D((2, 2)),
    Flatten(),
    Dense(10, activation='softmax')
])

In [8]:
# Компіляція моделі
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [9]:
# Навчання моделі
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 2ms/step - accuracy: 0.9035 - loss: 0.3233
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 2ms/step - accuracy: 0.9799 - loss: 0.0689
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9840 - loss: 0.0519
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9870 - loss: 0.0446
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9892 - loss: 0.0345


In [12]:
from sklearn.metrics import classification_report
# Оцінка моделі на тестових даних
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print('Test accuracy:', test_acc)

# Звіт про класифікацію
y_pred = model.predict(x_test)
y_pred_classes = np.argmax(y_pred, axis=1)
print(classification_report(y_test, y_pred_classes))

313/313 - 0s - 1ms/step - accuracy: 0.9820 - loss: 0.0561
Test accuracy: 0.9819999933242798
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       980
           1       0.99      0.99      0.99      1135
           2       0.99      0.96      0.97      1032
           3       0.98      0.99      0.98      1010
           4       0.99      0.99      0.99       982
           5       0.98      0.98      0.98       892
           6       0.99      0.98      0.99       958
           7       0.97      0.99      0.98      1028
           8       0.96      0.98      0.97       974
           9       0.98      0.97      0.98      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000



In [14]:
model.fit(x_train, y_train, epochs=10)

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9901 - loss: 0.0310
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9926 - loss: 0.0248
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.9940 - loss: 0.0192
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9944 - loss: 0.0183
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 5s 2ms/step - accuracy: 0.9958 - loss: 0.0144
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9963 - loss: 0.0121
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9972 - loss: 0.0097
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9971 - loss: 0.0092
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.9976 - loss: 0.0079
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9979 - loss: 0.0066


In [15]:
# Оцінка моделі на тестових даних
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=2)
print('Test accuracy:', test_acc)

# Звіт про класифікацію
y_pred = model.predict(x_test)
y_pred_classes = np.argmax(y_pred, axis=1)
print(classification_report(y_test, y_pred_classes))

313/313 - 1s - 2ms/step - accuracy: 0.9808 - loss: 0.0751
Test accuracy: 0.9807999730110168
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step
              precision    recall  f1-score   support

           0       0.99      0.98      0.98       980
           1       0.99      0.99      0.99      1135
           2       0.98      0.98      0.98      1032
           3       0.98      0.98      0.98      1010
           4       0.98      0.99      0.98       982
           5       0.97      0.99      0.98       892
           6       0.97      0.99      0.98       958
           7       0.98      0.97      0.98      1028
           8       0.99      0.97      0.98       974
           9       0.99      0.96      0.97      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000



Висновки:

MLP: Підходить для завдань з відносно простими даними, але коли йдеться про зображення, цей підхід не є оптимальним. Точність в районі 89-91% є непоганою для MLP, але ця мережа не може врахувати складні ознаки, що відображаються в зображеннях.

CNN (VGG16): Завдяки своїм властивостям CNN значно краще підходить для обробки зображень. Використання VGG16 дозволяє досягти точності понад 98% за рахунок вилучення ознак зображень, які важко вловити MLP.

Таким чином, згорткова нейромережа з VGG16 є більш ефективною для класифікації зображень, оскільки краще вловлює просторові та текстурні особливості зображень, що дозволяє досягти вищої точності, ніж MLP.






